presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCacheWriter.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.metadata;
+
+import com.facebook.presto.common.io.DataOutput;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.units.DataSize;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Accumulates data for DWRF stripe cache using given mode and max size.
+ */
+public class DwrfStripeCacheWriter
+{
+    private final DwrfStripeCacheMode mode;
+    private final DynamicSliceOutput cache;
+    private final long maxSize;
+    private final List<Integer> offsets = new ArrayList<>();
+
+    private int size;
+    private boolean canWriteMore = true;
+
+    public DwrfStripeCacheWriter(DwrfStripeCacheMode mode, DataSize maxSize)
+    {
+        // make sure max size is in the int range
+        this.maxSize = toIntExact(requireNonNull(maxSize, "maxSize is null").toBytes());
+        this.mode = requireNonNull(mode, "mode is null");
+        this.cache = new DynamicSliceOutput(64);
+        if (mode == DwrfStripeCacheMode.NONE) {
+            canWriteMore = false;
+        }
+    }
+
+    public void addIndexStreams(List<DataOutput> indexStreams, long indexSize)
+    {
+        if (!canWriteMore || !mode.hasIndex()) {
+            return;
+        }
+
+        if (size + indexSize > maxSize) {
+            canWriteMore = false;
+            return;
+        }
+
+        indexStreams.forEach(indexStream -> indexStream.writeData(cache));
```

Comment:
DynamicSliceOutput does not allow control over how the memory grows. We can't cap the memory at maxSize. 

So when we set the memory threshold to 8MB, stripe Cache Writer can consume 16MB buffer. Not sure is there an easy way to fix it though.

Reply:
I take majority of the *_max numbers as best effort numbers in terms of memory consumption. I don't think we should spend time making it super precise or enforcing the max size on the DynamicSliceOutput level.

I checked the DynamicSliceOutput, looks like it grows the slice by `newCapacity <<= 1` every time it reaches capacity. So if the max size is one of the powers of two then this situation should not happen, in worst case it would be the closest higher power of two.